### DISCUSSION TITLE
FastAPI receives headers by joining rather than duplicate(multiple) headers

### DISCUSSION THREAD

**Discussion by hard-coders** [fastapi/fastapi#6784] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from typing import List, Optional

from fastapi import FastAPI, Header

app = FastAPI()


@app.get("/items/")
async def read_items(x_token: Optional[List[str]] = Header(None)):

    return {"X-Token values": x_token}
```

### Description

* What I did with HTTPie(and same as what Swagger did with curl):
```bash
$ http -v :8000/items/ X-Token:first,second
GET /items/ HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: localhost:8000
User-Agent: HTTPie/2.3.0
X-Token: first,second
```
* It returns a JSON with 
```json
{
    "X-Token values": [
        "first,second"
    ]
}
```
* But I expected it to return:
```json
{
    "X-Token values": [
        "first",
        "second"
    ]
}
```

### Environment

* OS: macOS 11.2.3
* FastAPI Version: 0.63.0
* Python version: 3.9.4

### Additional context

From [Header Parameters docs](https://fastapi.tiangolo.com/tutorial/header-params/#duplicate-headers), it said FastAPI can receive multiple values for a single header. However, It's not working properly. Thus, I had to pop it and split it with `,`.

It's not a big deal, but I think the official doc must have reliability. Frankly, I'm confused this is a bug or a feature...

**Comment by four43** [comment, marked as answer]
You didn't send multiple headers like the doc example:
```
X-Token: foo
X-Token: bar
```

You sent one header with a comma in it.

**Comment by hard-coders** [comment]
@four43 Oops, What a shame...you are right. Thanks.

**Comment by four43** [comment]
That's a weird HTTP spec thing. Like with Accept it has multiple values and weights in there but it's a single header. It's weird. Glad it's working better for you.

**Comment by tiangolo** [comment]
Thanks for the help here @four43 ! 👏  🙇 

Thanks for reporting back and closing the issue @hard-coders  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by melenudo** [comment]
I know this is an old question, but FastAPI’s behavior is confusing. If you declare a header as a list and use the Swagger UI, you’ll get a comma-separated string. That is, from the UI, FastAPI won’t work as expected because Swagger will never send duplicate headers. This is the default Swagger serialization for header parameters: https://swagger.io/docs/specification/v3_0/serialization/
